common/lockless_queue.c

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/internal/lockless_queue.h>
+#include <stdlib.h>
+
+/* functions for lockless_queue_node */
+/*---------------------------------------------------------------------------*/
+void init_lockless_queue_node(lockless_queue_node* p_node)
+{
+    p_node->p_link = NULL;
+} /* init_lockless_queue_node */
+
+/* functions for lockless_queue */
+/*---------------------------------------------------------------------------*/
+void init_lockless_queue(lockless_queue* p_queue)
+{
+#ifdef _MSC_VER
+    InterlockedExchangePointer(&(p_queue->p_tail), NULL);
+    InterlockedExchangePointer(&(p_queue->p_head), NULL);
+#elif defined __GNUC__
+    __atomic_store_n(&(p_queue->p_tail), NULL, __ATOMIC_RELAXED);
+    __atomic_store_n(&(p_queue->p_head), NULL, __ATOMIC_RELAXED);
+#endif // __GNUC__
+} /* init_lockless_queue */
+
+void lockless_queue_push(lockless_queue* p_queue, lockless_queue_node* p_node)
+{
+    lockless_queue_node* p_expected = NULL;
+#ifdef _MSC_VER
+    lockless_queue_node* p_actual = NULL;
+#endif // _MSC_VER
+    do
+    {
+#ifdef _MSC_VER
+        p_expected = p_actual;
+        p_node->p_link = p_expected;
+        p_actual = (lockless_queue_node*)InterlockedCompareExchangePointer(
+            &(p_queue->p_tail), p_node, p_expected);
+    } while (p_actual != p_expected);
+#elif defined __GNUC__
+        p_node->p_link = p_expected;
+    } while (!__atomic_compare_exchange_n(
+        &(p_queue->p_tail),
+        &p_expected,
+        p_node,
+        1,
+        __ATOMIC_ACQ_REL,
+        __ATOMIC_ACQUIRE));
+#endif // __GNUC__
+} /* lockless_queue_push */
+
+lockless_queue_node* lockless_queue_pop(lockless_queue* p_queue)
```

Comment:
It is not obvious that this is a FIFO queue. STL style push_back/pop_front naming would help.